### PR TITLE
Fix Rails/HasManyOrHasOneDependent

### PR DESCRIPTION
**Pull request by macanudo527**
#### What? Why?

- Helps #6055 

This update makes dependent relationships explicit. Instead of the default behavior of `nil`, explicit actions have been declared for each relationship. This helps reduce database bulk by eliminating orphaned records. Also, in some cases, this helps to preserve user privacy by eliminating personal information connected to a user account if it is deleted. 

#### What should we test?
A lot of relationships have been updated in the database, enforcing the correct behaviour when records are deleted. So it would be worth testing what happens when they are each deleted. To test, they would also need to have the following dependent records present:

  * customer -> with orders (that is, try deleting a customer that has orders)
  * enterprise ->
    * producer properties
    * customers
    * inventory items
    * tag rules
  * order cycle ->
    * coordinator_fee_refs
    * cached_incoming_exchanges
    * cached_outgoing_exchanges
    * order_cycle_schedules
  * country -> states
  * credit card -> payment
  * order ->
    * state_changes
    * invoices (should not be allowed)
    * proxy_order
  * payment -> offsets
  * payment method (soft delete) -> credit_cards
  * property -> producer_properties
  * return authorisation (soft delete) -> inventory_units
  * shipment -> state_changes
  * shipping category -> 
    * products
    * shipping_method_categories
  * shipping method (soft delete) ->
    * shipping_rates
    * shipping_method_categories
    * distributor_shipping_methods
  * taxonomy -> taxons
  * user -> 
    * owned_enterprises
    * customers
    * credit_cards
  * variant (soft delete)
  * variant_overrides
  * inventory_items

#### Release notes

Changelog Category: Technical changes

### CODE REVIEW

File: app/models/customer.rb [hunk 1]
```diff
@@ -16,7 +16,7 @@ class Customer < ApplicationRecord
 
   belongs_to :enterprise
   belongs_to :user, class_name: "Spree::User", optional: true
-  has_many :orders, class_name: "Spree::Order"
+  has_many :orders, class_name: "Spree::Order", dependent: :destroy
```
Comment (dacook) [inline app/models/customer.rb:19]: Hmm this doesn't look right: I would think that we don't want to be destroying orders.

I had a quick look around and noticed at the bottom of this file the intention is to handle this situation differently:
https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/models/customer.rb#L73

So we want orders to stay when a customer is destroyed. Rails doesn't seem to have an explicit option like `dependent: :keep`, so I guess we need to disable the rubocop here. It can be done with an inline comment, I think like this from memory:

```suggestion
  has_many :orders, class_name: "Spree::Order" # rubocop:disable Rails/HasManyOrHasOneDependent
```

**Review by dacook — COMMENTED** [review]
Thanks Neal. 
I think each of these cases might require further consideration. I've looked into the first one which I think needs changing.

The others I'm not sure yet; I should be able to continue looking at these tomorrow.

File: app/models/customer.rb (hunk 1 above)
Comment (dacook) [reply app/models/customer.rb:19]: Ok I now realise that `dependent: :nullify` is what we want here; that's what the code I linked to is doing already.

File: app/models/enterprise.rb [hunk 2]
```diff
@@ -37,30 +37,31 @@ class Enterprise < ApplicationRecord
...
                                    class_name: 'EnterpriseGroup'
-  has_many :producer_properties, foreign_key: 'producer_id'
+  has_many :producer_properties, foreign_key: 'producer_id', dependent: :destroy
   has_many :properties, through: :producer_properties
   has_many :supplied_products, class_name: 'Spree::Product',
                                foreign_key: 'supplier_id',
                                dependent: :destroy
-  has_many :distributed_orders, class_name: 'Spree::Order', foreign_key: 'distributor_id'
+  has_many :distributed_orders, class_name: 'Spree::Order', foreign_key: 'distributor_id',
+                                dependent: :destroy
```
Comment (dacook) [inline app/models/enterprise.rb:46]: As per above, I don't think we want to delete orders. I think `:nullify` is what we want here.

File: app/models/enterprise.rb [hunk 3]
```diff
@@ -37,30 +37,31 @@ class Enterprise < ApplicationRecord
...
   belongs_to :address, class_name: 'Spree::Address'
   belongs_to :business_address, optional: true, class_name: 'Spree::Address', dependent: :destroy
-  has_many :enterprise_fees
+  has_many :enterprise_fees, dependent: :destroy
```
Comment (dacook) [inline app/models/enterprise.rb:49]: I see that the enterprise fees model `acts_as_paranoid`, which is a confusingly-named helper that implements soft-delete.
That is, we don't want to delete these, we want to keep them for reference (but hidden from view). So probably best to `:nullify` in this case.

File: app/models/spree/variant.rb [hunk 4]
```diff
@@ -38,7 +38,7 @@ class Variant < ApplicationRecord
 
     has_many :stock_items, dependent: :destroy, inverse_of: :variant
     has_many :stock_locations, through: :stock_items
-    has_many :stock_movements, dependent: :destroy
+    has_many :stock_movements, through: :stock_items, dependent: :destroy
```
Comment (dacook) [inline app/models/spree/variant.rb:41]: Was this broken before? That's interesting, perhaps we weren't using it and should consider removing it instead..

File: app/models/spree/tax_rate.rb [hunk 5]
```diff
@@ -21,7 +21,7 @@ class TaxRate < ApplicationRecord
 
     belongs_to :zone, class_name: "Spree::Zone", inverse_of: :tax_rates
     belongs_to :tax_category, class_name: "Spree::TaxCategory", inverse_of: :tax_rates
-    has_many :adjustments, as: :originator, dependent: :destroy
+    has_many :adjustments, as: :originator, dependent: nil
```
Comment (dacook) [inline app/models/spree/tax_rate.rb:24]: I guess this is acknowledging that the default behaviour hasn't been a problem thus far.. but it seems to me it would be better to `:nullify`. That might require some code changes, which I don't suggest we do now. 
I'd rather just leave this (and all below) as `todo` until we need to change it.

**Review by dacook — COMMENTED** [review]
This is a tricky one. I think it reveals a lot of gaps where we simply don't handle deletions very well at all. 
But they are apparently not problems at the moment, probably because we don't methods to delete these records anyway. Or perhaps we gracefully handle the broken links when it does happen.

I think it reveals that there's room for improvement, but I don't suggest we try to improve it for the sake of the linter. There's a few that I've commented on that we could try proceeding with, but I'd rather suggest that all the others are left as "todo" for now.

@mkllnk I'm wondering what you think about these?

File: app/models/customer.rb (hunk 1 above)
Comment (macanudo527) [reply app/models/customer.rb:19]: In order to use the default, which is what was there before, you can just add `dependent: nil`. That is always an option. I just got a little aggressive with db cleaning I guess.

File: app/models/spree/variant.rb (hunk 4 above)
Comment (macanudo527) [reply app/models/spree/variant.rb:41]: Yeah, the destroy wasn't working without this `through:`. A [quick search](https://github.com/search?q=repo%3Aopenfoodfoundation%2Fopenfoodnetwork%20stock_movements&type=code) seems to reveal that this relation is never used, it's only used with `stock_item`.

Should I just delete it?

Comment (dacook) [reply app/models/spree/variant.rb:41]: Thanks, yes it sounds like it!
FYI @mkllnk

**Comment by macanudo527** [comment]
> This is a tricky one. I think it reveals a lot of gaps where we simply don't handle deletions very well at all. But they are apparently not problems at the moment, probably because we don't methods to delete these records anyway. Or perhaps we gracefully handle the broken links when it does happen.
> 
> I think it reveals that there's room for improvement, but I don't suggest we try to improve it for the sake of the linter. There's a few that I've commented on that we could try proceeding with, but I'd rather suggest that all the others are left as "todo" for now.

When I started working on this I started to feel like this would be a tough one. We can leave these as TODOs, or just simply declare the default action explicitly `dependent: nil`. There is nothing wrong with the default really. It seems to have been working and that is what is needed for some of these relationships. If that changes, we can always modify the relationship later.

**Review by mkllnk — COMMENTED** [review]
This is a tough one. If we just wanted to comply with the style rule then we could just add `nil` everywhere but that doesn't improve anything. It's better to keep the todo list so that we know that we need to deal with it.

We have a few foreign keys that already say: `on_delete: :cascade`. This is in the schema file but it's not recorded in the models. I'm wondering if there's a good way to document it there. Rails shouldn't need to do anything there.

Some of our models are only soft-deleted. I'm not 100% sure but I would think that a `dependent: :destroy` is also just soft-deleting.

I don't have time to review it all today but I can have a look next week to see which changes seem simple. The rest probably needs individual pull requests. In some cases we probably want to remove code instead of styling it. And in some cases we may want to optimise the database and add cascading there to avoid callbacks. @macanudo527, do you think you are familiar enough with the code to decide about these or would it be better to leave it for someone else?

**Comment by macanudo527** [comment]
> Some of our models are only soft-deleted. I'm not 100% sure but I would think that a `dependent: :destroy` is also just soft-deleting.

`acts_as_paranoid` overrides destroy in order to soft delete. So `dependent: :destroy` just soft deletes the record. I think `delete` will actually delete the record though.

> I don't have time to review it all today but I can have a look next week to see which changes seem simple. The rest probably needs individual pull requests. In some cases we probably want to remove code instead of styling it. And in some cases we may want to optimise the database and add cascading there to avoid callbacks. @macanudo527, do you think you are familiar enough with the code to decide about these or would it be better to leave it for someone else?

I could give it a try. I might have some questions about the logic, though. Should I work on inforcing the foreign keys first? They may be needed to help with cascading deletes in the database and help me visualize the data a little better.

**Comment by mkllnk** [comment]
> acts_as_paranoid overrides destroy in order to soft delete. So dependent: :destroy just soft deletes the record. I think delete will actually delete the record though.

That's an additional complication. It means that we could have `dependent: destroy` but if the record is only soft-deleted then it can still hold a foreign key which will prevent it referenced record from being deleted.

We actually have a few records like enterprises which we cannot delete because we don't handle these cases. And we never specified how we would like to solve it. I guess I'm trying to tell you not to expect too much because you probably won't be able to resolve all cases.

File: app/models/spree/address.rb [hunk 6]
```diff
@@ -13,7 +13,7 @@ class Address < ApplicationRecord
     belongs_to :state, class_name: "Spree::State"
 
     has_one :enterprise, dependent: :restrict_with_exception
-    has_many :shipments
+    has_many :shipments, dependent: :nullify
```
Comment (mkllnk) [inline app/models/spree/address.rb:16]: I think that this is an invalid case. You need to assign a new address to a shipment before you can delete the old one. I'm But usually we update addresses in place and keep the id. I wonder sometimes how many orphan addresses we have because that model is re-used in so many places. Too many places, actually, but that is going off topic here.

Hm, the database and Rails actually allow a shipment without address. Not sure if it causes server problems down the line though... anyway, I guess we never do this.

File: app/models/spree/payment.rb [hunk 7]
```diff
@@ -21,7 +21,7 @@ class Payment < ApplicationRecord
     belongs_to :payment_method, class_name: 'Spree::PaymentMethod'
 
     has_many :offsets, -> { where("source_type = 'Spree::Payment' AND amount < 0").completed },
-             class_name: "Spree::Payment", foreign_key: :source_id
+             class_name: "Spree::Payment", foreign_key: :source_id, dependent: :nullify
```
Comment (mkllnk) [inline app/models/spree/payment.rb:24]: How confusing is this. It looks like there are payments which have another payment as source. They call it an offset here. I guess that it's the same as a (partial) refund.

This is now saying that if we delete a payment, we are keeping the its offsets. Same as before but it doesn't make much sense to me. Maybe good to make it explicit now though and not reference a non-existing record.

File: app/models/spree/return_authorization.rb [hunk 8]
```diff
@@ -8,8 +8,8 @@ class ReturnAuthorization < ApplicationRecord
 
     belongs_to :order, class_name: 'Spree::Order', inverse_of: :return_authorizations
 
-    has_many :inventory_units, inverse_of: :return_authorization
-    has_one :stock_location
+    has_many :inventory_units, inverse_of: :return_authorization, dependent: :nullify
+    has_one :stock_location, dependent: :nullify
```
Comment (mkllnk) [inline app/models/spree/return_authorization.rb:12]: Not sure if this has any impact, probably not.

File: app/models/spree/user.rb [hunk 9]
```diff
@@ -35,9 +35,9 @@ class User < ApplicationRecord
     has_many :enterprise_roles, dependent: :destroy
     has_many :enterprises, through: :enterprise_roles
     has_many :owned_enterprises, class_name: 'Enterprise',
-                                 foreign_key: :owner_id, inverse_of: :owner
+                                 foreign_key: :owner_id, inverse_of: :owner, dependent: :nullify
     has_many :owned_groups, class_name: 'EnterpriseGroup',
-                            foreign_key: :owner_id, inverse_of: :owner
+                            foreign_key: :owner_id, inverse_of: :owner, dependent: :nullify
```
Comment (mkllnk) [inline app/models/spree/user.rb]: Enterprises and groups must have an owner. I think that we should raise when trying to delete here. That's probably already happening because the foreign key constraint is violated. Nullifying would allow us to delete these users and have orphaned, invalid enterprises in the database.

**Review by mkllnk — CHANGES_REQUESTED** [review]
Thank you for raising this. The impacts can be a bit difficult to predict though. I think I gave some indicators in my comments. I agree with @dacook that it's better to keep some of these as _todo_ and not defined `nil` because that could be seen as intentional.

Where from here? Probably reduce this PR to the easy and undisputed cases and then you can open pull requests for other cases that need more thought.

File: app/models/spree/stock_item.rb [hunk 10]
```diff
@@ -8,7 +8,7 @@ class StockItem < ApplicationRecord
 
     belongs_to :stock_location, class_name: 'Spree::StockLocation', inverse_of: :stock_items
     belongs_to :variant, -> { with_deleted }, class_name: 'Spree::Variant'
-    has_many :stock_movements, dependent: :destroy
+    has_many :stock_movements, dependent: :delete_all # delete_all required to avoid ReadOnlyError
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/spree/stock_item.rb:11]: I'll revert and mark this as TODO. I opened an issue #11506 .

**Comment by macanudo527** [comment]
@mkllnk Should I create separate issues for each one of these or one general one?

File: app/models/customer.rb [hunk 11]
```diff
@@ -16,7 +16,7 @@ class Customer < ApplicationRecord
 
   belongs_to :enterprise
   belongs_to :user, class_name: "Spree::User", optional: true
-  has_many :orders, class_name: "Spree::Order"
+  has_many :orders, class_name: "Spree::Order", dependent: :nullify
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/customer.rb:19]: Keeping

File: app/models/enterprise.rb [hunk 12]
```diff
@@ -37,30 +37,31 @@ class Enterprise < ApplicationRecord
                                     dependent: :destroy
   has_and_belongs_to_many :groups, join_table: 'enterprise_groups_enterprises',
                                    class_name: 'EnterpriseGroup'
-  has_many :producer_properties, foreign_key: 'producer_id'
+  has_many :producer_properties, foreign_key: 'producer_id', dependent: :destroy
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/enterprise.rb:40]: Keeping

File: app/models/enterprise.rb [hunk 13]
```diff
@@ -37,30 +37,31 @@ class Enterprise < ApplicationRecord
...
   has_many :properties, through: :producer_properties
   has_many :supplied_products, class_name: 'Spree::Product',
                                foreign_key: 'supplier_id',
                                dependent: :destroy
-  has_many :distributed_orders, class_name: 'Spree::Order', foreign_key: 'distributor_id'
+  has_many :distributed_orders, class_name: 'Spree::Order', foreign_key: 'distributor_id',
+                                dependent: :nullify
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/enterprise.rb:46]: I'll revert and mark it as TODO.

File: app/models/enterprise.rb [hunk 14]
```diff
@@ -37,30 +37,31 @@ class Enterprise < ApplicationRecord
...
   belongs_to :address, class_name: 'Spree::Address'
   belongs_to :business_address, optional: true, class_name: 'Spree::Address', dependent: :destroy
-  has_many :enterprise_fees
+  has_many :enterprise_fees, dependent: :nullify
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/enterprise.rb:49]: Should I add a `:restrict_with_error` or `:restrict_with_exception` to this (for now)? or just revert and TODO?

File: app/models/enterprise.rb [hunk 15]
```diff
@@ -37,30 +37,31 @@ class Enterprise < ApplicationRecord
...
   belongs_to :owner, class_name: 'Spree::User',
                      inverse_of: :owned_enterprises
   has_many :distributor_payment_methods,
-           inverse_of: :distributor, foreign_key: :distributor_id
+           inverse_of: :distributor, foreign_key: :distributor_id, dependent: :destroy
   has_many :distributor_shipping_methods,
-           inverse_of: :distributor, foreign_key: :distributor_id
+           inverse_of: :distributor, foreign_key: :distributor_id, dependent: :destroy
   has_many :payment_methods, through: :distributor_payment_methods
   has_many :shipping_methods, through: :distributor_shipping_methods
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/enterprise.rb:59]: I'll revert and mark as TODO.

File: app/models/enterprise.rb [hunk 16]
```diff
@@ -37,30 +37,31 @@ class Enterprise < ApplicationRecord
...
-  has_many :customers
-  has_many :inventory_items
-  has_many :tag_rules
+  has_many :customers, dependent: :destroy
+  has_many :inventory_items, dependent: :destroy
+  has_many :tag_rules, dependent: :destroy
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/enterprise.rb:62]: Keeping

File: app/models/enterprise.rb [hunk 17]
```diff
@@ -37,30 +37,31 @@ class Enterprise < ApplicationRecord
...
   has_one :stripe_account, dependent: :destroy
-  has_many :vouchers
+  has_many :vouchers, dependent: :destroy
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/enterprise.rb:64]: I'll revert and mark as TODO

File: app/models/spree/order.rb [hunk 18]
```diff
@@ -71,12 +71,12 @@ def states
              },
              class_name: 'Spree::Adjustment',
              dependent: :destroy
-    has_many :invoices
+    has_many :invoices, dependent: :destroy
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/spree/order.rb:74]: mark this with `:restrict_with_exception`?

File: app/models/spree/payment_method.rb [hunk 19]
```diff
@@ -15,7 +15,7 @@ class PaymentMethod < ApplicationRecord
     DISPLAY = [:both, :back_end].freeze
     default_scope -> { where(deleted_at: nil) }
 
-    has_many :credit_cards, class_name: "Spree::CreditCard"
+    has_many :credit_cards, class_name: "Spree::CreditCard", dependent: :destroy
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/spree/payment_method.rb:18]: keeping

File: app/models/spree/property.rb [hunk 20]
```diff
@@ -4,7 +4,7 @@ module Spree
   class Property < ApplicationRecord
     has_many :product_properties, dependent: :destroy
     has_many :products, through: :product_properties
-    has_many :producer_properties
+    has_many :producer_properties, dependent: :destroy
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/spree/property.rb:7]: keeping

File: app/models/spree/shipment.rb [hunk 21]
```diff
@@ -12,7 +12,7 @@ class Shipment < ApplicationRecord
 
     has_many :shipping_rates, dependent: :delete_all
     has_many :shipping_methods, through: :shipping_rates
-    has_many :state_changes, as: :stateful
+    has_many :state_changes, as: :stateful, dependent: :destroy
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/spree/shipment.rb:15]: keeping?

File: app/models/spree/shipping_category.rb [hunk 22]
```diff
@@ -4,7 +4,7 @@ module Spree
   class ShippingCategory < ApplicationRecord
     validates :name, presence: true
     has_many :products
-    has_many :shipping_method_categories, inverse_of: :shipping_method
+    has_many :shipping_method_categories, inverse_of: :shipping_method, dependent: :destroy
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/spree/shipping_category.rb:7]: OK, created a new issue for this #11507

File: app/models/spree/variant.rb [hunk 23]
```diff
@@ -33,12 +33,12 @@ class Variant < ApplicationRecord
 
     delegate_belongs_to :product, :name, :description, :meta_keywords
 
-    has_many :inventory_units, inverse_of: :variant
-    has_many :line_items, inverse_of: :variant
+    has_many :inventory_units, inverse_of: :variant, dependent: :destroy
+    has_many :line_items, inverse_of: :variant, dependent: :destroy
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/spree/variant.rb]: I'll revert and mark as TODO.

File: app/models/spree/variant.rb [hunk 24]
```diff
@@ -54,10 +54,10 @@ class Variant < ApplicationRecord
     delegate_belongs_to :default_price, :display_price, :display_amount,
                         :price, :price=, :currency
 
-    has_many :exchange_variants
+    has_many :exchange_variants, dependent: :destroy
```
[the remark this answers is not in the document]
Comment (macanudo527) [reply app/models/spree/variant.rb:57]: Maybe. I'm just wondering why the model needs to 'reload':
```
    def destruction
      exchange_variants.reload.destroy_all
      yield
    end
```

Tests pass if I comment out the callback and leave the default behavior for the `dependent` relationship, so this is not covered, at least in `variant_spec.rb`. There are no specs for `exchange_variant`

File: app/models/spree/address.rb (hunk 6 above)
Comment (macanudo527) [reply app/models/spree/address.rb:16]: I'll revert this and mark it as TODO for now. It seems like there should be a better solution for this though. For example, we could soft-delete the address to keep it. Rewriting the address seems like we are losing something.

File: app/models/spree/payment.rb (hunk 7 above)
Comment (macanudo527) [reply app/models/spree/payment.rb:24]: keeping

File: app/models/spree/return_authorization.rb (hunk 8 above)
Comment (macanudo527) [reply app/models/spree/return_authorization.rb:12]: keeping

File: app/models/spree/user.rb (hunk 9 above)
Comment (macanudo527) [reply app/models/spree/user.rb]: changing to :restrict_with_exception

File: app/models/enterprise.rb (hunk 14 above)
[the remark this answers is not in the document]
Comment (mkllnk) [reply app/models/enterprise.rb:49]: The closest to the current behaviour is `:restrict_with_exception` because foreign keys raise an exception at the moment. But since we don't quite know how to handle this case, just leave it as it was before.

File: app/models/spree/order.rb (hunk 18 above)
[the remark this answers is not in the document]
Comment (mkllnk) [reply app/models/spree/order.rb:74]: Yes, exception sounds right here.

File: app/models/spree/shipment.rb (hunk 21 above)
[the remark this answers is not in the document]
Comment (mkllnk) [reply app/models/spree/shipment.rb:15]: Yes, I guess so. The only downside would be _if_ we ever made shipments soft-deletable and then lose state information. But that sounds very unlikely. More performance sounds better in the meantime. The less data the better.

File: app/models/spree/variant.rb (hunk 24 above)
[the remark this answers is not in the document]
Comment (mkllnk) [reply app/models/spree/variant.rb:57]: I'm not sure. Let's keep the old behaviour for now until someone has time to look at it properly.

File: app/models/spree/order.rb (hunk 18 above)
[the remark this answers is not in the document]
Comment (mkllnk) [reply app/models/spree/order.rb:74]: This still needs restricting.

File: app/models/spree/shipping_method.rb [hunk 25]
```diff
@@ -13,11 +13,11 @@ class ShippingMethod < ApplicationRecord
...
     default_scope -> { where(deleted_at: nil) }
 
-    has_many :shipping_rates, inverse_of: :shipping_method
+    has_many :shipping_rates, inverse_of: :shipping_method, dependent: :destroy
     has_many :shipments, through: :shipping_rates
-    has_many :shipping_method_categories
-    has_many :shipping_categories, through: :shipping_method_categories
-    has_many :distributor_shipping_methods
+    has_many :shipping_method_categories, dependent: :destroy
+    has_many :shipping_categories, through: :shipping_method_categories, dependent: :destroy
```
Comment (mkllnk) [inline app/models/spree/shipping_method.rb:19]: No, we don't want to delete categories when deleting shipping methods.

File: app/models/spree/user.rb [hunk 26]
```diff
@@ -14,15 +14,15 @@ class User < ApplicationRecord
...
+    has_many :orders, dependent: nil
     belongs_to :ship_address, class_name: 'Spree::Address'
     belongs_to :bill_address, class_name: 'Spree::Address'
 
     has_and_belongs_to_many :spree_roles,
                             join_table: 'spree_roles_users',
                             class_name: "Spree::Role"
 
-    has_many :spree_orders, class_name: "Spree::Order"
+    has_many :spree_orders, class_name: "Spree::Order", dependent: nil
```
Comment (mkllnk) [inline app/models/spree/user.rb:25]: I never noticed this. `orders` and `spree_orders` is the same relationship. I'll open a PR to remove the latter...

**Review by mkllnk — CHANGES_REQUESTED** [review]
Almost. :smile:

File: app/models/spree/shipping_method.rb (hunk 25 above)
Comment (macanudo527) [reply app/models/spree/shipping_method.rb:19]: My understanding is that this code will destroy `shipping_method_categories` (ie. the record that links the two) only and not the actual record in `shipping_categories` when using a `through:` relationship like this, [according to the API documentation](https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#:~:text=have%20been%20deleted.-,What%20gets%20deleted%3F,-There%20is%20a).

Comment (mkllnk) [reply app/models/spree/shipping_method.rb:19]: Well, you have associations to both models here. You can destroy shipping_method_categories and not destroy shipping_categories. That's what I meant.

**Review by mkllnk — APPROVED** [review]

File: app/models/spree/shipping_method.rb [hunk 27]
```diff
@@ -13,11 +13,11 @@ class ShippingMethod < ApplicationRecord
...
     default_scope -> { where(deleted_at: nil) }
 
-    has_many :shipping_rates, inverse_of: :shipping_method
+    has_many :shipping_rates, inverse_of: :shipping_method, dependent: :destroy
     has_many :shipments, through: :shipping_rates
-    has_many :shipping_method_categories
-    has_many :shipping_categories, through: :shipping_method_categories
-    has_many :distributor_shipping_methods
+    has_many :shipping_method_categories, dependent: :destroy
+    has_many :shipping_categories, through: :shipping_method_categories, dependent: nil
```
Comment (dacook) [inline app/models/spree/shipping_method.rb:19]: It seems to be that the `dependent` option is irrelevant for `has_many through:`  associations.
We've already specified above that the join table record should get destroyed, and the `shipping_categories` record doesn't have a foreign key in it. 

I tried removing this option and Rubocop doesn't complain. This is consistent with the below `has_many through` association.

```suggestion
    has_many :shipping_categories, through: :shipping_method_categories
```

File: app/models/spree/variant.rb (hunk 4 above)
Comment (dacook) [reply app/models/spree/variant.rb:41]: I think we agreed to delete association entirely, as it's unused. The cleanup  is done by the `has_many :stock_items` association, so we don't need to specify anything here.
```suggestion
```

**Review by dacook — CHANGES_REQUESTED** [review]
I have a suggestion to change one of the `dependent` choices, and a couple of cleanups

@mkllnk or @macanudo527 are you able to confirm each of these?

**Comment by macanudo527** [comment]
@dacook Thanks for the check. I fixed those issues. It should be good to go now.

**Review by dacook — APPROVED** [review]
Great, thanks Neal!

**Comment by dacook** [comment]
Ok so I didn't notice the testing details haven't been filled out yet. To work out how to test, had to poke around a bit and randomly tested what happens when deleting a return authorisation. It broke!

I've fixed that one now, but it causes enough alarm for me to suggest we should test all updated dependent options, which is a lot...

**Comment by dacook** [comment]
I've updated the testing notes.
@mkllnk can you think of any way around doing all that testing?

**Comment by mkllnk** [comment]
> can you think of any way around doing all that testing?

No. And it may not be obvious to the tester how to delete certain things. Like deleting an exchange means removing a product from an order cycle.

Some of this may have been broken beforehand as well. So we need to test before and after to spot the difference if any.

**Comment by mkllnk** [comment]
@filipefurtad0 What do you think about testing this? Could you flag the associations you don't know how to test and then we can fill in details?

**Comment by filipefurtad0** [comment]
> @filipefurtad0 What do you think about testing this? Could you flag the associations you don't know how to test and then we can fill in details?

Sure, I'm up for it - thanks for pinging. Just to be sure, you mean the usual manual testing process or is it useful to add specs if/where needed? (like in e54852f739302efe510aef696173329d0aadb56c.)

**Comment by mkllnk** [comment]
Manual testing.

**Comment by filipefurtad0** [comment]
Hey @macanudo527 and team,

This looks good. I've tried to stick to the test cases below, and not unfold them too much, otherwise this could easily loose scope. It got a bit gnarly, especially as there are some items which I could not really find a way to delete, through the UI (or the UI made it just complicated, like deleting enterprises... if there is a big list). Anyway, all in all, I think we're good:

- customer -> :heavy_check_mark: 
- enterprise -> there is no error message, but the enterprise cannot deleted
        producer properties
        customers
        inventory items
        tag rules
- order cycle ->
        coordinator_fee_refs -  :heavy_check_mark: 
        cached_incoming_exchanges - not sure how to test deleting cached incoming exchanges... I've created them in the UI and, before updating the DB, opened up a parallel tab and deleted the OC; I think this should clear cached exchanges  :heavy_check_mark: 
        cached_outgoing_exchanges - same as above, all good.  :heavy_check_mark: 
        order_cycle_schedules - no error, but OC is not deleted. This is the same behavior as in master.  :heavy_check_mark: 
- country -> states . not sure one can delete countries from the UI? At least I could not find a way.
- credit card -> payment - OK. Deleted a saved credit card after making a payment with it. All good. :heavy_check_mark: 
- order -> I guess one cannot delete orders... right?
        state_changes -> tested state changes, all as before.  :heavy_check_mark: 
        invoices (should not be allowed) - not sure what to test here? 
        proxy_order - also not sure?
- payment ->  - can we delete payments? I think we can only change the state.
- payment method (soft delete) -> credit_cards - OK; payment is soft deleted without issues, and cannot be foudn or edited even as superadmin  :heavy_check_mark: 
- property -> producer_properties - OK; after creating and attributing a property to a producer, deleting it (superadmin) removes the attribution to the producer  :heavy_check_mark: 
- return authorisation (soft delete) -> inventory_units - I could delete return authorizattions 
- shipment -> state_changes - OK; shipment changes from ready to shipped, as expected
- shipping category -> can't delete shipping categories, if associated to products. this is the current behaviour as well, on master: https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/652037504ce7af0008dfa4c7?filters[error.status]=open&filters[event.since]=30d&event_id=652326b200c59f299f4f0000
        products - error 500 (above)
        shipping_method_categories - OK
- shipping method (soft delete) -> OK
        shipping_rates
        shipping_method_categories
        distributor_shipping_methods
- taxonomy -> taxons - OK, no error 500. However it disappears from the taxonomuy list as superadmin, but is still visible on the product category dropdown, when editing a product.
- user -> none works, error 500; all pre-existing
        owned_enterprises 
        customers
        credit_cards
- variant (soft delete) - OK, for ordered products
    variant_overrides - deleted product with variant override - OK
    inventory_items - deleted product with inventory item - OK


PS: Pic from UI change when deleting a return authorization / authorization not deleted - error 401:
image

Is there a specific scenario which should be investigated in more depth?

**Comment by mkllnk** [comment]
Great, thank you. I don't think that there's any specific case that needs special attention. I'll merge this then.

**Review by mkllnk — APPROVED** [review]

**Comment by filipefurtad0** [comment]
> deleting a return authorization / authorization not deleted - error 401:

I think this might be a regression, but I'd say an edge case. I'll re-test in other servers after staging master and open an issue if verified.

**Comment by macanudo527** [comment]
@filipefurtad0 Thanks for doing all the legwork on the tests! 

@mkllnk Thanks for all the feedback and help! I learned a lot.
